tests/utils.go

```diff
@@ -5105,3 +5106,37 @@ func VMILauncherIgnoreWarnings(virtClient kubecli.KubevirtClient) func(vmi *v1.V
 		return vmi
 	}
 }
+
+func CreateAndWaitForVmi(client kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
+	vmi, err := client.VirtualMachineInstance(vmi.Namespace).Create(vmi)
+	if err != nil {
+		return nil, err
+	}
+
+	vmi = WaitUntilVMIReady(vmi, libnet.WithIPv6(console.LoginToCirros))
+	return vmi, nil
+}
+
+func DeleteVmiAndWaitForDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) error {
+	if vmi != nil {
+		err := virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+
+		err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
```

Comment:
You can simply write `Eventually(ThisVMI(vmi), 120).Should(BeGone())`

Reply:
You can write `EventuallyWithOffset(1, ThisVMI(vmi), 120).Should(BeGone())` to see the error at the right location in the log traceback. I can understand that you want to return an error, but I would prefer to get rid of all the custom loop implementations and consolidate on the Ginkgo provided one.

Reply:
`libvmi` is a helper package that may be used by different callers at different stack level calls.
Using the offset will not support such calls (from different stack levels) and we saw this is causing a mess in the error trace.

Therefore, I think it is most preferable not to assert inside helpers and let this be done at the test level. This is also much more transparent to a test reader.
Assert helpers by themselves are fine, but then we are not mixing other operations with the assert (e.g. `assertVMIExists`).

Reply:
I see a few options here, but all the custom wait logic is a problem:

 1) Whenever possible, just doing the explicit `Eventually` in the test. With the advanced matchers they are pretty expressive. The waitfor helpers are not very explicit, one always has to check exactly what they are waiting for and they vary how they are implemented. We have seen issues with them too.
 2) If helpers are preferred to have e.g. a `EventuallyVMIShouldBeGone(vmi)` which can use the `Eventually inside` and maybe accompany it with a `EventuallyVMIShouldBeGoneWithOffset(1, vmi).
3) I don't mind if some deep-down helpers would return an error for some things to be easier concatenated, but in general what we do in all test helpers is just checking if we reach a goal and just fail the test without any special cleanup or countermeasures, which makes the error return pretty redundant.

Reply:
> The waitfor helpers are not very explicit, one always has to check exactly what they are waiting for and they vary how they are implemented. We have seen issues with them too.

If they are implemented correctly, with established methods (in this case, `PollImmediate` from k8s), I think we can trust them.

But I am fine with helpers which do not retry internally, they just return responses and at the test level these are retried with the ginkgo tooling. I just think this is not always possible or efficient.

Option 2, to me, is just saying: `assertVMIShouldBeGoneWithOffset(1, vmi)`, which is fine, it just suffers from the inability to share this code out of this project (i.e. we couple everything to ginkgo, which is not a good strategy IMO).
Having helper packages depending on the ginkgo framework feels wrong to me in general and avoiding it will put better boundaries.
Said that, this is a good middle ground we could probably come to an agreement.

> but in general what we do in all test helpers is just checking if we reach a goal and just fail the test without any special cleanup or countermeasures, which makes the error return pretty redundant.

I am aware of this pattern. I think it is not an ideal one, suffering from things like:
- The offset issues (we do not control who calls the helpers and from what stack "distance").
- Inability to add informative info on the way "back" (e.g. add to the error relevant info if something fails).
- Inability to cleanup on a per-test or per-action. This is a framework limitation, which became a reason to work differently in kubevirt.
- Coupling to a specific framework. Which makes it hard to share the helpers with other projects or to be flexible and replace the framework if a better one comes to be.

For me, just the "offset" is a good enough reason to not embed asserts in actionable helpers. And having the version with the offset as an argument is just pure ugly IMO.

Reply:
IIUC option 2 is kind of what we have right now: 
https://github.com/kubevirt/kubevirt/blob/ae8cd8e6a66f711431c46eed31d0f6ccb3117bc6/tests/libvmi/operations.go#L38-L40

Or did I miss something?

Reply:
At the end the `WaitForVmiDeletion` is not needed and directly placing the `Eventually` in the code should be good enough. I don't think the rest needs to be discussed any further.

Reply:
Done. 
What I did was to remove the helpers completely, because they didn't have much use, and the only place where we wait for the vmi deleting I returned the `Eventually` implementation.